api/notifications/notifications.go

```diff
@@ -0,0 +1,202 @@
+package notifications
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/gorilla/websocket"
+
+	"github.com/Peripli/service-manager/pkg/query"
+	"github.com/Peripli/service-manager/pkg/types"
+	"github.com/Peripli/service-manager/pkg/util"
+	"github.com/Peripli/service-manager/storage"
+
+	"github.com/Peripli/service-manager/pkg/log"
+
+	"github.com/Peripli/service-manager/pkg/web"
+	"github.com/Peripli/service-manager/pkg/ws"
+)
+
+const (
+	lastKnownRevisionHeader     = "last_known_revision"
+	lastKnownRevisionQueryParam = "last_known_revision"
+)
+
+func (c *Controller) handleWS(req *web.Request) (*web.Response, error) {
+	user, _ := web.UserFromContext(req.Context())
+	notificationQueue, lastKnownRevision, err := c.notificator.RegisterConsumer(user)
+	if err != nil {
+		return nil, fmt.Errorf("could not register notification consumer: %v", err)
+	}
+
+	revisionKnownToProxy := 0
+	revisionKnownToProxyStr := req.URL.Query().Get(lastKnownRevisionQueryParam)
+	if revisionKnownToProxyStr != "" {
+		revisionKnownToProxy, err = strconv.Atoi(revisionKnownToProxyStr)
+		if err != nil {
+			c.unregisterConsumer(notificationQueue)
```

Comment:
Is it possible to add these in defer somehow? It might be forgotten if this code is refactored in the future.

Reply:
Not sure if this is possible, a test is better to catch this kind of behavior. I will try to add a test for this case